NPE: NetworkUtils' context not initialized

```
java.lang.RuntimeException: Unable to start receiver de.danoeh.antennapod.receiver.ConnectivityActionReceiver: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.Object android.content.Context.getSystemService(java.lang.String)' on a null object reference
	at android.app.ActivityThread.handleReceiver(ActivityThread.java:2649)
	at android.app.ActivityThread.access$1800(ActivityThread.java:154)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1398)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loop(Looper.java:135)
	at android.app.ActivityThread.main(ActivityThread.java:5294)
	at java.lang.reflect.Method.invoke(Native Method)
	at java.lang.reflect.Method.invoke(Method.java:372)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:904)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:699)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.Object android.content.Context.getSystemService(java.lang.String)' on a null object reference
	at de.danoeh.antennapod.core.util.NetworkUtils.autodownloadNetworkAvailable(NetworkUtils.java:47)
	at de.danoeh.antennapod.receiver.ConnectivityActionReceiver.onReceive(ConnectivityActionReceiver.java:23)
	at android.app.ActivityThread.handleReceiver(ActivityThread.java:2642)
	... 9 more
```

--- a/app/src/main/java/de/danoeh/antennapod/PodcastApp.java
+++ b/app/src/main/java/de/danoeh/antennapod/PodcastApp.java
@@ -8,11 +8,8 @@ import com.joanzapata.iconify.Iconify;
 import com.joanzapata.iconify.fonts.FontAwesomeModule;
 import com.joanzapata.iconify.fonts.MaterialModule;
 
+import de.danoeh.antennapod.core.ClientConfig;
 import de.danoeh.antennapod.core.feed.EventDistributor;
-import de.danoeh.antennapod.core.preferences.PlaybackPreferences;
-import de.danoeh.antennapod.core.preferences.UserPreferences;
-import de.danoeh.antennapod.core.storage.PodDBAdapter;
-import de.danoeh.antennapod.core.util.NetworkUtils;
 import de.danoeh.antennapod.spa.SPAUtil;
 
 /** Main application class. */
@@ -56,11 +53,8 @@ public class PodcastApp extends Application {
 
 		singleton = this;
 
-		PodDBAdapter.init(this);
-		UserPreferences.init(this);
-		UpdateManager.init(this);
-		PlaybackPreferences.init(this);
-		NetworkUtils.init(this);
+		ClientConfig.initialize(this);
+
 		EventDistributor.getInstance();
 		Iconify.with(new FontAwesomeModule());
 		Iconify.with(new MaterialModule());

--- a/app/src/main/java/de/danoeh/antennapod/receiver/ConnectivityActionReceiver.java
+++ b/app/src/main/java/de/danoeh/antennapod/receiver/ConnectivityActionReceiver.java
@@ -8,6 +8,7 @@ import android.net.NetworkInfo;
 import android.text.TextUtils;
 import android.util.Log;
 
+import de.danoeh.antennapod.core.ClientConfig;
 import de.danoeh.antennapod.core.storage.DBTasks;
 import de.danoeh.antennapod.core.storage.DownloadRequester;
 import de.danoeh.antennapod.core.util.NetworkUtils;
@@ -20,7 +21,8 @@ public class ConnectivityActionReceiver extends BroadcastReceiver {
 		if (TextUtils.equals(intent.getAction(), ConnectivityManager.CONNECTIVITY_ACTION)) {
 			Log.d(TAG, "Received intent");
 
-			if (NetworkUtils.autodownloadNetworkAvailable()) {
+            ClientConfig.initialize(context);
+            if (NetworkUtils.autodownloadNetworkAvailable()) {
 				Log.d(TAG, "auto-dl network available, starting auto-download");
 					DBTasks.autodownloadUndownloadedItems(context);
 			} else { // if new network is Wi-Fi, finish ongoing downloads,

--- a/app/src/main/java/de/danoeh/antennapod/receiver/PowerConnectionReceiver.java
+++ b/app/src/main/java/de/danoeh/antennapod/receiver/PowerConnectionReceiver.java
@@ -3,10 +3,9 @@ package de.danoeh.antennapod.receiver;
 import android.content.BroadcastReceiver;
 import android.content.Context;
 import android.content.Intent;
-import android.os.BatteryManager;
 import android.util.Log;
 
-import de.danoeh.antennapod.core.BuildConfig;
+import de.danoeh.antennapod.core.ClientConfig;
 import de.danoeh.antennapod.core.preferences.UserPreferences;
 import de.danoeh.antennapod.core.storage.DBTasks;
 import de.danoeh.antennapod.core.storage.DownloadRequester;
@@ -25,6 +24,7 @@ public class PowerConnectionReceiver extends BroadcastReceiver {
 
         Log.d(TAG, "charging intent: " + action);
 
+        ClientConfig.initialize(context);
         if (Intent.ACTION_POWER_CONNECTED.equals(action)) {
             Log.d(TAG, "charging, starting auto-download");
             // we're plugged in, this is a great time to auto-download if everything else is

--- a/app/src/main/java/de/danoeh/antennapod/receiver/SPAReceiver.java
+++ b/app/src/main/java/de/danoeh/antennapod/receiver/SPAReceiver.java
@@ -9,8 +9,8 @@ import android.widget.Toast;
 
 import java.util.Arrays;
 
-import de.danoeh.antennapod.BuildConfig;
 import de.danoeh.antennapod.R;
+import de.danoeh.antennapod.core.ClientConfig;
 import de.danoeh.antennapod.core.feed.Feed;
 import de.danoeh.antennapod.core.storage.DownloadRequestException;
 import de.danoeh.antennapod.core.storage.DownloadRequester;
@@ -27,30 +27,30 @@ public class SPAReceiver extends BroadcastReceiver{
 
     @Override
     public void onReceive(Context context, Intent intent) {
-        if (TextUtils.equals(intent.getAction(), ACTION_SP_APPS_QUERY_FEEDS_REPSONSE)) {
-            if (BuildConfig.DEBUG) Log.d(TAG, "Received SP_APPS_QUERY_RESPONSE");
-            if (intent.hasExtra(ACTION_SP_APPS_QUERY_FEEDS_REPSONSE_FEEDS_EXTRA)) {
-                String[] feedUrls = intent.getStringArrayExtra(ACTION_SP_APPS_QUERY_FEEDS_REPSONSE_FEEDS_EXTRA);
-                if (feedUrls != null) {
-                    if (BuildConfig.DEBUG) Log.d(TAG, "Received feeds list: " + Arrays.toString(feedUrls));
-                    for (String url : feedUrls) {
-                        Feed f  = new Feed(url, null);
-                        try {
-                            DownloadRequester.getInstance().downloadFeed(context, f);
-                        } catch (DownloadRequestException e) {
-                            Log.e(TAG, "Error while trying to add feed " + url);
-                            e.printStackTrace();
-                        }
-                    }
-                    Toast.makeText(context, R.string.sp_apps_importing_feeds_msg, Toast.LENGTH_LONG).show();
-
-                } else {
-                    Log.e(TAG, "Received invalid SP_APPS_QUERY_REPSONSE: extra was null");
-                }
-            } else {
-                Log.e(TAG, "Received invalid SP_APPS_QUERY_RESPONSE: Contains no extra");
+        if (!TextUtils.equals(intent.getAction(), ACTION_SP_APPS_QUERY_FEEDS_REPSONSE)) {
+            return;
+        }
+        Log.d(TAG, "Received SP_APPS_QUERY_RESPONSE");
+        if (!intent.hasExtra(ACTION_SP_APPS_QUERY_FEEDS_REPSONSE_FEEDS_EXTRA)) {
+            Log.e(TAG, "Received invalid SP_APPS_QUERY_RESPONSE: Contains no extra");
+            return;
+        }
+        String[] feedUrls = intent.getStringArrayExtra(ACTION_SP_APPS_QUERY_FEEDS_REPSONSE_FEEDS_EXTRA);
+        if (feedUrls == null) {
+            Log.e(TAG, "Received invalid SP_APPS_QUERY_REPSONSE: extra was null");
+            return;
+        }
+        Log.d(TAG, "Received feeds list: " + Arrays.toString(feedUrls));
         ClientConfig.initialize(context);
+        for (String url : feedUrls) {
+            Feed f = new Feed(url, null);
+            try {
+                DownloadRequester.getInstance().downloadFeed(context, f);
+            } catch (DownloadRequestException e) {
+                Log.e(TAG, "Error while trying to add feed " + url);
+                e.printStackTrace();
             }
         }
+        Toast.makeText(context, R.string.sp_apps_importing_feeds_msg, Toast.LENGTH_LONG).show();
     }
 }

--- a/app/src/main/java/de/danoeh/antennapod/receiver/SPAReceiver.java
+++ b/app/src/main/java/de/danoeh/antennapod/receiver/SPAReceiver.java
@@ -49,6 +49,7 @@ public class SPAReceiver extends BroadcastReceiver{
                 }
             } else {
                 Log.e(TAG, "Received invalid SP_APPS_QUERY_RESPONSE: Contains no extra");
+        ClientConfig.initialize(context);
             }
         }
     }

--- a/core/src/main/java/de/danoeh/antennapod/core/ClientConfig.java
+++ b/core/src/main/java/de/danoeh/antennapod/core/ClientConfig.java
@@ -1,5 +1,12 @@
 package de.danoeh.antennapod.core;
 
+import android.content.Context;
+
+import de.danoeh.antennapod.core.preferences.PlaybackPreferences;
+import de.danoeh.antennapod.core.preferences.UserPreferences;
+import de.danoeh.antennapod.core.storage.PodDBAdapter;
+import de.danoeh.antennapod.core.util.NetworkUtils;
+
 /**
  * Stores callbacks for core classes like Services, DB classes etc. and other configuration variables.
  * Apps using the core module of AntennaPod should register implementations of all interfaces here.
@@ -22,4 +29,19 @@ public class ClientConfig {
     public static FlattrCallbacks flattrCallbacks;
 
     public static DBTasksCallbacks dbTasksCallbacks;
+
+    private static boolean initialized = false;
+
+    public static synchronized void initialize(Context context) {
+        if(initialized) {
+            return;
+        }
+        PodDBAdapter.init(context);
+        UserPreferences.init(context);
+        UpdateManager.init(context);
+        PlaybackPreferences.init(context);
+        NetworkUtils.init(context);
+        initialized = true;
+    }
+
 }

--- a/core/src/main/java/de/danoeh/antennapod/core/UpdateManager.java
+++ b/core/src/main/java/de/danoeh/antennapod/core/UpdateManager.java
@@ -1,4 +1,4 @@
-package de.danoeh.antennapod;
+package de.danoeh.antennapod.core;
 
 
 import android.content.Context;

--- a/core/src/main/java/de/danoeh/antennapod/core/receiver/AlarmUpdateReceiver.java
+++ b/core/src/main/java/de/danoeh/antennapod/core/receiver/AlarmUpdateReceiver.java
@@ -6,7 +6,7 @@ import android.content.Intent;
 import android.text.TextUtils;
 import android.util.Log;
 
-import de.danoeh.antennapod.core.preferences.PlaybackPreferences;
+import de.danoeh.antennapod.core.ClientConfig;
 import de.danoeh.antennapod.core.preferences.UserPreferences;
 
 /** Listens for events that make it necessary to reset the update alarm. */
@@ -22,8 +22,7 @@ public class AlarmUpdateReceiver extends BroadcastReceiver {
 		} else if (TextUtils.equals(intent.getAction(), Intent.ACTION_PACKAGE_REPLACED)) {
 			Log.d(TAG, "Resetting update alarm after app upgrade");
 		}
-        PlaybackPreferences.init(context);
-        UserPreferences.init(context);
+		ClientConfig.initialize(context);
         UserPreferences.restartUpdateAlarm(false);
 	}
 

--- a/core/src/main/java/de/danoeh/antennapod/core/receiver/FeedUpdateReceiver.java
+++ b/core/src/main/java/de/danoeh/antennapod/core/receiver/FeedUpdateReceiver.java
@@ -5,6 +5,7 @@ import android.content.Context;
 import android.content.Intent;
 import android.util.Log;
 
+import de.danoeh.antennapod.core.ClientConfig;
 import de.danoeh.antennapod.core.preferences.UserPreferences;
 import de.danoeh.antennapod.core.storage.DBTasks;
 import de.danoeh.antennapod.core.util.NetworkUtils;
@@ -19,6 +20,7 @@ public class FeedUpdateReceiver extends BroadcastReceiver {
     @Override
     public void onReceive(Context context, Intent intent) {
         Log.d(TAG, "Received intent");
+        ClientConfig.initialize(context);
         if (NetworkUtils.isDownloadAllowed()) {
             DBTasks.refreshAllFeeds(context, null);
         } else {

--- a/core/src/main/java/de/danoeh/antennapod/core/receiver/MediaButtonReceiver.java
+++ b/core/src/main/java/de/danoeh/antennapod/core/receiver/MediaButtonReceiver.java
@@ -6,6 +6,7 @@ import android.content.Intent;
 import android.util.Log;
 import android.view.KeyEvent;
 import de.danoeh.antennapod.core.BuildConfig;
+import de.danoeh.antennapod.core.ClientConfig;
 import de.danoeh.antennapod.core.service.playback.PlaybackService;
 
 /** Receives media button events. */
@@ -21,6 +22,7 @@ public class MediaButtonReceiver extends BroadcastReceiver {
 		KeyEvent event = (KeyEvent) intent.getExtras().get(
 				Intent.EXTRA_KEY_EVENT);
 		if (event.getAction() == KeyEvent.ACTION_DOWN) {
+			ClientConfig.initialize(context);
 			Intent serviceIntent = new Intent(context, PlaybackService.class);
 			int keycode = event.getKeyCode();
 			serviceIntent.putExtra(EXTRA_KEYCODE, keycode);

--- a/core/src/main/java/de/danoeh/antennapod/core/receiver/MediaButtonReceiver.java
+++ b/core/src/main/java/de/danoeh/antennapod/core/receiver/MediaButtonReceiver.java
@@ -5,7 +5,7 @@ import android.content.Context;
 import android.content.Intent;
 import android.util.Log;
 import android.view.KeyEvent;
-import de.danoeh.antennapod.core.BuildConfig;
+
 import de.danoeh.antennapod.core.ClientConfig;
 import de.danoeh.antennapod.core.service.playback.PlaybackService;
 
@@ -18,9 +18,8 @@ public class MediaButtonReceiver extends BroadcastReceiver {
 
 	@Override
 	public void onReceive(Context context, Intent intent) {
-		if (BuildConfig.DEBUG) Log.d(TAG, "Received intent");
-		KeyEvent event = (KeyEvent) intent.getExtras().get(
-				Intent.EXTRA_KEY_EVENT);
+		Log.d(TAG, "Received intent");
+		KeyEvent event = (KeyEvent) intent.getExtras().get(Intent.EXTRA_KEY_EVENT);
 		if (event.getAction() == KeyEvent.ACTION_DOWN) {
 			ClientConfig.initialize(context);
 			Intent serviceIntent = new Intent(context, PlaybackService.class);

